### ISSUE TITLE
Introduce `#[builder(overwritable)]`

### ISSUE THREAD

**Issue by Veetaha** [elastio/bon#149]
This new attribute can be placed on individual members or at the top level via `#[builder(on(_, overwritable)]`. It disables the compile-time check for overwriting already set members. For example with this attribute this code compiles:

```rust
#[derive(bon::Builder)]
struct Example {
    #[builder(overwritable)]
    x: u32,
}

Example::builder()
    .x(1)
    .x(2)
    .build();
```

Why is this useful? It removes a whole bun of generic state variables from the generated code and considerably improves the compile times. For example in the `frankenstein` repo the difference for me is around 1.5-2 seconds of compile time.

@EdJoPaTo would you be interested in enabling this in `frankenstein` to improve the compile times (even at the cost of allowing overwrites)? To me it looks like the problem of overwrites is not that big and the compilation time perf. gains are probably more important.

I have this feature already implemented on my branch https://github.com/elastio/bon/pull/145, and I'm planning to include it in the next release.

**Comment by EdJoPaTo** [comment]
The name itself doesn't include something in relation to performance currently so this might be misleading…

Thought about having this as a feature or something to enable it only on debug or something like that. But it doesn't seem that good either…

Not sure what I think about mixing feature and performance thingy… 🤔

**Comment by Veetaha** [comment]
Well, it can't be a cargo feature because it changes the type signature of the builder (it removes some generic associated types from the builder's state trait). I'm planning to make that type signature stable (i.e. to be able to name the actual type of the builder). So if anyone enables it as a cargo feature, their code may cease to compile if downstream code depends on some type states being available.

Therefore, this should be configured via attributes. The name of the attribute indeed doesn't make it clear that it improves the compilation performance, but I'm don't think that's important, because the attribute name describes what it actually does (makes it possible to call the setter multiple times for the same member). The fact, that it simplifies the generics machinery under the hood is just a considerable bonus, which will be part of the docs (and maybe have a separate page on "improving compilation perf" in the guide). I expect people will just enable this option in a single place (like with an attribute alias), and put a descriptive comment that say "it improves the compile perf. by N seconds).

**Comment by Veetaha** [comment]
So, you'd still like to keep the overwrites disabled? My expectation was that it wouldn't be as important since the overwrites problem doesn't happen that often, and people are used to this when working with manually-written builders. For some, it may even be a feature

**Comment by EdJoPaTo** [comment]
im not entirely sure about it. Setting something twice seems like a mistake so having that is a code smell. Normally code smells are something for the linters… Could it create another kind of warning?

Also… is this a requirement for the basic library? Maybe it should allow overrides all the time and have a flag to prevent setting it more than once? Then all the people get better compile time performance for free and only the ones that need the override prevention can enable it? Maybe that's a better default?

**Comment by Veetaha** [comment]
Well, the difference between `overwritable` and non-`overwritable` becomes visible only at a greater scale (especially in structs with 50+ fields). In the general case, for structs of reasonable sizes and counts, having the overwrites protection makes sense by default, so I this option would only be useful for projects of bigger scale like `frankenstein`. This is where you already use the attribute alias approach, whereby requiring users to have an attribute alias by default to protect from overwrites doesn't seem reasonable.

Unfortunately, there isn't a good way to write a custom lint in Rust today. There are some tools that do that, but they are unpopular and 99% of people don't use anything fancier than `clippy`. 

> Setting something twice seems like a mistake so having that is a code smell

Yes, but another scenario could be where the developer uses this feature to implement a base fixture for tests. For example, the tests need to fill only the fields they are interested in and keep all other fields with dummy values. 

```rust

#[derive(bon::Builder)]
#[cfg_attr(test, builder(on(_, overwritable)))]
struct User {
    login: String,
    name: String,
    level: u32,
}

#[cfg(test)]
mod tests {
    use super::*;
    use user_builder::{SetLevel, SetLogin, SetName};

    fn user() -> UserBuilder<SetLevel<SetName<SetLogin>>> {
        User::builder()
            .level(24)
            .name("Bon Bon".to_owned())
            .login("@bonbon".to_owned())
    }

    #[test]
    fn test_with_empty_name() {
        // Build user with an empty name, and all other fields filled with dummy data
        let user = user()
            .name("".to_owned())
            .build();

        // ...
    }

    #[test]
    fn test_with_zero_level_and_admin_login() {
        let admin = user()
            .level(0)
            .login("admin".to_owned())
            .build();

        // ...
    }
}
```

I think it's actually a really handy (and might also deserve a separate guide page).

**Comment by Veetaha** [comment]
Yeah, I see you points. Anyway, I already have this feature implemented in my branch. I'll probably just keep it disabled by default and make it available only if the user enables some `experimental` cargo feature flag waiting for some more user feedback on the use cases.

**Comment by EdJoPaTo** [comment]
Is there also a plan / time to remove the feature when it doesn't seem to be used and no one stated their use case? Keeping the code clean seems like a good idea :innocent:

**Comment by Veetaha** [comment]
I don't think this particular feature bears high maintenance burden. I'll just keep it experimental for 6 months, for example. And if I receive no feedback on this issue, I'll remove it

**Comment by MuhannadAlrusayni** [comment]
Hello, glad to see you have implemented this feature already! Thank you!

For my use case, I am building UI structs that looks like this:
```rust
#[derive(bon::Builder)]
#[builder(on(String, into))]
pub struct TextInput {
    name: String,
    value: Option<String>,
    error: Option<String>,
}
```
And let's say the `value` and `error` values comes from multiple sources and they have precedences too. With these requirements I would write something like this:

```rust
// let's say the value is an email (e.g. example@domain.com)
let value_from_db = db_get_current_value().ok();
let value_from_user_input = user_input_value();
let default_value = String::from("default value");

let error_from_user_input = value_from_user_input.as_ref().and_then(|email| email.validate().err());
let system_error = value_from_user_input.as_ref().and_then(|email| verify_email_is_available(email).err());

// Building the UI
let email_input = TextInput::builder()
    .name("email")
    .value(default_value)
    .maybe_value(value_from_db)
    .maybe_value(value_from_user_input)
    .maybe_error(error_from_user_input)
    .maybe_error(system_error)
    .build()
```
Note that I didn't test the snippet, it's just for illustration.

I already use TypedBuilder for this, other types that doesn't require this and mutators, I use bon for it.

Hope this help.

**Comment by Veetaha** [comment]
> I already use TypedBuilder for this

Could you elaborate how you configured `TypedBuilder` to allow for overwrites? As I know, `TypedBuilder` doesn't allow calling setters repeatedly, and there is no attribute to disable the overwrites protection.

Anyway, I think the code you shown contains a mistake. Every call to a setter overwrites the previous value. For example, if you do smth like `.a(Some(1)).a(None)`, then the resulting value in `a` will be `None`. There is no special behavior for `Option<T>` in this case.

Just think of this like this (with the `.value()/maybe_value()` from your example):

```rust
let  mut value = Some(default_value); // equivalent to `value(default_value)`
value = value_from_db;                // equivalent to `maybe_value(value_from_db)`
value = value_from_user_input;        // equivalent to `value_from_user_input`
```
So it's more like assigning to a variable. I think from the code above it's clear that the last assignment fully overwrites previous assignments.

Instead, if you want to have a chain of defaults, then you should use `Option::or`, `Option::or_else` and `Option::unwrap_or` methods to implement that.

```rust
#[derive(bon::Builder)]
#[builder(on(String, into))]
pub struct TextInput {
    name: String,
    value: String,
    error: Option<String>,
}

let value_from_db: Option<String> = todo!();
let value_from_user_input: Option<String> = todo!();
let default_value = String::from("default value");

let error_from_user_input: Option<String> = todo!();
let system_error: Option<String> = todo!();

// Building the UI
let email_input = TextInput::builder()
    .name("email")
    .value(
        // This chain will stop at the first non-None value,
        // or on the final `unwrap_or`
        value_from_user_input
            .or(value_from_db)
            .unwrap_or(default_value),
    )
    .maybe_error(error_from_user_input.or(system_error))
    .build();
```

If you do it like this, you'll also notice that `value` should not be wrapped in `Option<...>`, because it always has a default value, so it can never be `None`.

**Comment by Veetaha** [comment]
@MuhannadAlrusayni 

> I already use TypedBuilder for this

Could you elaborate how you configured `TypedBuilder` to allow for overwrites? As I know, `TypedBuilder` doesn't allow calling setters repeatedly, and there is no attribute to disable the overwrites protection.

Anyway, I think the code you shown contains a mistake. Every call to a setter overwrites the previous value. For example, if you do smth like `.a(Some(1)).a(None)`, then the resulting value in `a` will be `None`. There is no special behavior for `Option<T>` in this case.

Just think of this like this (with the `.value()/maybe_value()` from your example):

```rust
let  mut value = Some(default_value); // equivalent to `value(default_value)`
value = value_from_db;                // equivalent to `maybe_value(value_from_db)`
value = value_from_user_input;        // equivalent to `value_from_user_input`
```
So it's more like assigning to a variable. I think from the code above it's clear that the last assignment fully overwrites previous assignments.

Instead, if you want to have a chain of defaults, then you should use `Option::or`, `Option::or_else` and `Option::unwrap_or` methods to implement that.

```rust
#[derive(bon::Builder)]
#[builder(on(String, into))]
pub struct TextInput {
    name: String,
    value: String,
    error: Option<String>,
}

let value_from_db: Option<String> = todo!();
let value_from_user_input: Option<String> = todo!();
let default_value = String::from("default value");

let error_from_user_input: Option<String> = todo!();
let system_error: Option<String> = todo!();

// Building the UI
let email_input = TextInput::builder()
    .name("email")
    .value(
        // This chain will stop at the first non-None value,
        // or on the final `unwrap_or`
        value_from_user_input
            .or(value_from_db)
            .unwrap_or(default_value),
    )
    .maybe_error(error_from_user_input.or(system_error))
    .build();
```

If you do it like this, you'll also notice that `value` should not be wrapped in `Option<...>`, because it always has a default value, so it can never be `None`.

**Comment by MuhannadAlrusayni** [comment]
Oh! sorry I forget that I have used mutators like this:
```rust
#[derive(TypedBuilder)]
#[builder(field_defaults(setter(into)), mutators(
    pub fn error(&mut self, error: impl ToString) {
        self.error = Some(value.to_string())
    }
    pub fn try_error(&mut self, error: Option<impl ToString>) {
        if let Some(error) = error {
            self.error = Some(error.to_string())
        }
    }
    pub fn value(&mut self, value: impl ToString) {
        self.value = Some(value.to_string())
    }
    pub fn try_value(&mut self, value: Option<impl ToString>) {
        if let Some(value) = value {
            self.value = Some(value.to_string())
        }
    }
))]
pub struct TextInput {
    name: String,
    value: Option<String>,
    error: Option<String>,
}
```

And yes, this is is not the same behaviour as overwrite would be, it's like `set_if_some` and I am not sure any more if it fits within this feature, but it's kind of related to it some how, since it needs the builder to allow set the value multiple times.

**Comment by Veetaha** [comment]
@MuhannadAlrusayni

Aha, with that `TypedBuilder` code, now it's clear. Indeed, I think this use case doesn't fit into the `overwritable` attribute.

If you'd like to adapt that code to `bon`, you'll need to use custom fields ([here is the guide](https://bon-rs.com/guide/typestate-api/custom-fields)).

The difference is that your custom setters need to be defined in a separate impl block for the builder:

```rust
#[derive(bon::Builder)]
pub struct TextInput {
    #[builder(field)]
    value: Option<String>,

    #[builder(field)]
    error: Option<String>,

    name: String,
}

impl<S: text_input_builder::State> TextInputBuilder<S> {
    pub fn error(&mut self, error: impl ToString) {
        self.error = Some(error.to_string())
    }
    pub fn try_error(&mut self, error: Option<impl ToString>) {
        if let Some(error) = error {
            self.error = Some(error.to_string())
        }
    }
    pub fn value(&mut self, value: impl ToString) {
        self.value = Some(value.to_string())
    }
    pub fn try_value(&mut self, value: Option<impl ToString>) {
        if let Some(value) = value {
            self.value = Some(value.to_string())
        }
    }
}
```

**Comment by angelorendina** [comment]
An issue I have with the `overwritable` feature is that it fiddles with the typestate.

A setter currently maps `Builder<S> -> Builder<SetField<S>>`, regardless if `S` already was `IsSet<Field>` or not. Arguably, I would expect the setter to only generate a different typestate if the field had not been previously assigned, something like
```
setter: Builder<S> -> Builder<SetField<S>> where S: IsUnset<Field>
setter: Builder<S> -> Builder<S> where S: IsSet<Field>
```

Also, as mentioned above by EdJoPaTo, perhaps a setter should only be allowed (at most) once, which would make `overwritable` a double edge sword.

I have fiddled a bit trying to implement `getter(mut)` and I got it working locally (I could tidy it up a bit and open a PR for that). My personal opinion is that setters should be once-only calls, and if you need to overwrite any field, you could do it with a `getter(mut)`.

**Comment by Veetaha** [comment]
@angelorendina, thank you for looking into this and considering a contribution!

`overwritable` does remove the `where S: IsUnset` bound by design. Although, how do you think this can be a problem for `getter(mut)`? I've recently found a somewhat obscure use case for `overwritable` when changing from a couple of `bool` setters to a single `enum` setter (https://github.com/kivikakk/comrak/pull/497#issuecomment-2521343524).

Also, could you share your use case for `getter(mut)`? I wonder if that use case could be satisfied with `#[builder(field)]` instead?

Feel free to message me at https://bon-rs.com/discord for easier communication

**Comment by angelorendina** [comment]
@Veetaha thanks for pointing out the `builder(field)` attribute: it _almost_ fits our use case.

I can actually repurpose the example in the documentation to explain our issue: imagine we needed the `Command` to have a non-empty list of arguments!

I have a workaround that looks like this:
```rust
#[derive(bon::Builder)]
struct Command {
    #[builder(field)]
    inner_args: Vec<String>,
    // marker field that at least one argument was passed, and the list is not empty
    args: (),
    name: String,
}

impl Command {
    fn to_string(self) -> String {
        // safe to assume inner_args is non-empty
        assert!(!self.inner_args.is_empty());
        todo!("should construct the string `name arg1 args2 ...`")
    }
}

impl<S> CommandBuilder<S> where S: command_builder::State, S::Args: command_builder::IsUnset<> {
    pub fn arg(self, arg: String) -> CommandBuilder<command_builder::SetArgs<S>> {
        self.args(()).append_arg(arg)
    }
}

impl<S> CommandBuilder<S> where S: command_builder::State, S::Args: command_builder::IsSet<> {
    // (*) Does not compile, because rustc cannot infer that this is mutually exclusive with `arg` above...
    // pub fn arg(mut self, arg: String) -> CommandBuilder<S> {
    //     self.inner_args.push(arg);
    //     self
    // }
}

impl<S> CommandBuilder<S> where S: command_builder::State, S::Args: command_builder::IsSet<> {
    // (*) ...so we have to give it a different name...
    pub fn append_arg(mut self, arg: String) -> CommandBuilder<S> {
        self.inner_args.push(arg);
        self
    }
}

#[test]
fn requires_arguments() {
    let command = Command::builder()
        .name("name".to_string())
        .arg("first".to_string())
        // (*) ...hence this does not compile...
        // .arg("would be nice".to_string())
        // (*) ...but this works fine.
        .append_arg("second".to_string())
        .build();

    // Does not compile since `args` is never set :)
    // let _no_compile = Command::builder().build();
}
```
but it would be nice if this kind of functionality was supported by a simple
```rust
#[derive(bon::Builder)]
struct Command {
    #[builder(field)]
    inner_args: Vec1<String>,
    name: String,
}
```

**Comment by Veetaha** [comment]
> I have a workaround that looks like this:

Thank you for sharing this. Although I don't see how the example use case you shown could be nicely solved with `getter(mut)`. I expect that `getter(mut)` will have `where S::Member: IsSet` bound on it, so you'd need to have an eagerly initialized member.

Also, here is a bit simpler workaround without `getter(mut)` that uses `overwritable` in this case. It solves the problem of `arg/append_arg` naming conflict:

```rust
pub struct Command {
    name: String,
    args: Vec<String>,
}

#[bon::bon]
impl Command {
    #[builder]
    fn new(
        #[builder(field)] inner_args: Vec<String>,

        #[builder(overwritable, setters(vis = ""))] _args: (),

        name: String,
    ) -> Self {
        Self {
            name,
            args: inner_args,
        }
    }
}

use command_builder::{SetArgs, State};

impl<S: State> CommandBuilder<S> {
    pub fn arg(mut self, arg: String) -> CommandBuilder<SetArgs<S>> {
        self.inner_args.push(arg);
        self.args(())
    }
}

fn main() {
    Command::builder()
        .name("bon".to_owned())
        .arg("one".to_owned())
        .arg("two".to_owned())
        .build();

    // This doesn't compile:
    // .build();
    //  ^^^^^ the member `Unset<args>` was not set, but this method requires it to be set
    Command::builder()
        .name("bon".to_owned())
        .build();
}
```

Do you have a specific example with `getter(mut)` that would solve the problem of `arg/append_arg` naming and a separate `#[builder(field)]`?

**Comment by Veetaha** [comment]
@angelorendina, I've just discovered a problem in the current `#[builder(overwritable)]`, which isn't critical, but still a bit ugly. The problem is that if you call the setter several times for the required member called `x1`, for example, then you'll end up with the builder's type state `SetX1<SetX1<SetX1>>`, when instead it'd be cool if the type state didn't change after the first call to the setter, and once the member is already set, no new `SetX1<>` state transition is added to the type.

By itself this nesting of `SetX1` on every setter call isn't bad on the surface. It's only a minor problem for the case when people want to spell the exact builder type.

However, I tried to work around this problem and suggest an alternative solution to the one proposed above (https://github.com/elastio/bon/issues/149#issuecomment-2523782192). With this approach the building process is split into two stages:

1. Fill all named parameters other than the `args`.
2. Fill `args` (here we can't intermix other named parameters).

As you can see, the limitation here is that the order of the building matter. Although, this may be a plus, because I don't think you'd want to see the pattern like `.arg("foo").name("bar").arg("bar")` where the `name` setter appears in the middle of `arg(...)` setters call chain.

```rust
use bon::{bon, builder};

pub struct Command {
    args: Vec<String>,
    name: String,
}

#[bon]
impl Command {
    // The first stage. Here we specify all regular named parameters.
    // We specify `arg` as the `finish_fn` parameter, and make the `finish_fn` named
    // after it. This way the `finish_fn` appears just like a regular setter to the caller,
    // however, it returns a new type of the builder, once it is called, thus transitioning
    // to a new state at the type level.
    #[builder(finish_fn(name = arg))]
    pub fn new(
        #[builder(finish_fn)] //
        arg: String,
        name: String,
    ) -> CommandWithArg1Builder {
        Self::with_arg1(arg, name)
    }

    // Here we accept all values from the stage 1, and add the ability to push new args
    // via `#[builder(field)]`. A small problem here is that `arg1` needs to be passed as
    // a `start_fn` parameter separate from `builder(field)`. Ideally, we'd have the ability
    // to access start_fn parameters just like regular `#[builder(field)]`s
    #[builder(builder_type(vis = "pub"), finish_fn = build)]
    fn with_arg1(
        #[builder(start_fn)] //
        arg1: String,

        #[builder(start_fn)] //
        name: String,

        #[builder(field)] //
        mut args: Vec<String>,
    ) -> Command {
        // That's suboptimal, but we can make `start_fn` parameters accessible in `self`
        // of the builder just like `builder(field)` params. We don't even need a getter(mut)
        // for that.
        args.insert(0, arg1);

        Command { args, name }
    }
}

impl<S: command_with_arg1_builder::State> CommandWithArg1Builder<S> {
    pub fn arg(mut self, arg: String) -> Self {
        self.args.push(arg);
        self
    }
}

fn main() {
    Command::builder()
        .name("echo".to_owned())
        .arg("Hello, World!".to_owned())
        .arg("asd".to_owned())
        .arg("asdsa".to_owned())
        .build();

    // Doesn't compile (method `build` was not found)
    // Command::builder()
    //     .name("echo".to_owned())
    //     .build();
}
```

Also, I think it's possible to swap the stages 1 and 2, if you'd like the caller to fill `args` first (not last).

So I think what would be cool here, is the ability to access `start_fn` as fields on the builder. This would make this pattern nicer.

**Comment by angelorendina** [comment]
@Veetaha indeed, the issue we are facing is that we do need to explicitly spell out the type of the builder: we have a state machine that wraps various stages of the builder, so we need each stage to explicitly state which fields are set or unset.

The issue you mentioned about overwritable is exactly what I meant when I originally said that I would expect
```rust
setter: Builder<S> -> Builder<SetField<S>> where S: IsUnset<Field>
setter: Builder<S> -> Builder<S> where S: IsSet<Field>
```
so that the setter avoid appending further `SetField` to the type, if it's already set.

I was fiddling with another possible implementation of the internals of the derive, and I got the following to work in a way that does not append multiple `SetField`. I am not sure how it compares to the internals of `bon`, but perhaps it could be repurposed or adapted:

```rust
/// The actual structure we will build.
struct Thing {
    a: u8,
    b: u16,
}

/// Marker for a field that was populated with a value
struct Set<T>(T);
/// Marker for a field that was not populated
struct Unset;

/// Only implemented for [Set] and [Unset]
trait Field {}
impl<T> Field for Set<T> {}
impl Field for Unset {}

/// Represents the state of each stage of the builder,
/// where its associated types correspond to the final fields
/// and are marked as either [Set] or [Unset].
trait State {
    type A: Field;
    type B: Field;
}

/// Marker for a builder whose fields have not been set.
struct Empty;
impl State for Empty {
    type A = Unset;
    type B = Unset;
}

/// Marker for a builder whose A field has been set.
struct WithA<S: State>(PhantomData<S>);
impl<S: State> State for WithA<S> {
    type A = Set<u8>;
    type B = S::B;
}

/// Marker for a builder whose B field has been set.
struct WithB<S: State>(PhantomData<S>);
impl<S: State> State for WithB<S> {
    type A = S::A;
    type B = Set<u16>;
}

/// The actual builder, together with its generic representing its completion state.
struct Builder<S: State> {
    a: S::A,
    b: S::B,
}

// (*) Wrapper workaround to allow for mutually exclusive implementations, because...
trait SetterA<S: State>: Sized {
    /// Sets the value of `a` for the builder.
    fn a(self, a: u8) -> Builder<S>;
}
impl<S: State<A = Unset>> SetterA<WithA<S>> for Builder<S> {
    // (*) ...this...
    fn a(self, a: u8) -> Builder<WithA<S>> {
        Builder {
            a: Set(a),
            b: self.b,
        }
    }
}
impl<S: State<A = Set<u8>>> SetterA<S> for Builder<S> {
    // (*) ...and this should be mutually exclusive,
    // but cause a compiler error due to a known bug.
    fn a(self, a: u8) -> Builder<S> {
        Builder {
            a: Set(a),
            b: self.b,
        }
    }
}

trait SetterB<S: State>: Sized {
    /// Sets the value of `b` for the builder.
    fn b(self, b: u16) -> Builder<S>;
}
impl<S: State<B = Unset>> SetterB<WithB<S>> for Builder<S> {
    fn b(self, b: u16) -> Builder<WithB<S>> {
        Builder {
            a: self.a,
            b: Set(b),
        }
    }
}
impl<S: State<B = Set<u16>>> SetterB<S> for Builder<S> {
    fn b(self, b: u16) -> Builder<S> {
        Builder {
            a: self.a,
            b: Set(b),
        }
    }
}

impl<S: State<A = Set<u8>, B = Set<u16>>> Builder<S> {
    /// Finishes the build, returning the fully populated [Thing].
    // Only available when all fields have been set, guaranteed at compile time by the trait bounds.
    fn build(self) -> Thing {
        Thing {
            a: self.a.0,
            b: self.b.0,
        }
    }
}

impl Thing {
    /// Gets a Builder for [Thing].
    fn builder() -> Builder<Empty> {
        Builder { a: Unset, b: Unset }
    }
}

#[test]
fn foo() {
    let x = Thing::builder().a(8).b(16).build();
    assert_eq!(x.a, 8);
    assert_eq!(x.b, 16);

    let y = Thing::builder().b(16).a(8).build();
    assert_eq!(y.a, 8);
    assert_eq!(y.b, 16);

    let z = Thing::builder()
        // appends WithA
        .a(8)
        // appends WithB
        .b(16)
        // does NOT append WithA
        .a(7)
        // does NOT append WithB
        .b(15)
        .build();
    assert_eq!(z.a, 7);
    assert_eq!(z.b, 15);

    // let _no_compile = Thing::builder()
    //     .a(8)
    //     .build();
}
```

**Comment by Veetaha** [comment]
> The issue you mentioned about overwritable is exactly what I meant

Yeah, now I see, I somehow missed that point initially 😳, sorry for that.

> I got the following to work

Thank you for sharing this, that's very helpful! I tried to fiddle with this approach as well, and adapt it to bon's codegen. However, the main problem that I see here is that this requires generating additional traits (one per each member), several method implementations (two per each member), which incurs compile time overhead. My initial idea for `overwritable` was that it would reduce compile time overhead (not increase it 😅).

However, even if there weren't any compile time overhead, the bigger problem for me is that the generated setters become trait methods instead of being inherent methods. This would require the user of the builder to import the necessary traits into their scope, which is inconvenient. It also increases the API surface of the builder, making it prone to more API breakages.

I tried to fiddle with the provided code, but I couldn't find a way to keep the setters as inherent methods (yet?). I'll try to fiddle with this more. Hopefully there is a better way to do that. Otherwise I think this may be blocked by https://github.com/rust-lang/rust/issues/20400

**Comment by dj8yfo** [comment]
actually, i'd find it quite usable, as it adds some flexibility, which is simply missing otherwise, and bon cannot be used at all

**Comment by Veetaha** [comment]
@dj8yfo could you share an example where `overwritable` is critical in your case?

**Comment by dj8yfo** [comment]
@Veetaha it'll pop up some time later in a link to this issue. Please don't mind if it appears silly to you

**Comment by XBagon** [comment]
I was looking for a way to reuse a builder, which currently doesn't work as all the `build` functions consume the builder.
Then I stumbled upon `overwritable` and realized it is probably needed to make reusable builders really useful.

My case: I want to build a set of slightly varying objects with a single builder I preconfigure/have a default for, and then just adjust a single field or so for each object. 
In some cases I might be able to clone the builder (with an overhead depending on its size), but in my current use-case I have refs in the builder I use to build objects cloning things from the refs. 
So for now I'll need to create a new builder for each object.

**Comment by Veetaha** [comment]
@XBagon could you share some example code for your use case?

Maybe defining a separate builder wrapper based on function with the defaults would fit your use case:
 
```
#[bon::builder]
fn builder_with_defaults(
    #[builder(default = 42)]
    a: u32,
    #[builder(default = 2 * a)]
    b: u32,
) -> T {
    main_builder().a(a).b(b).build()
}

// Now you can call the builder overriding the defaults as needed
builder_with_defaults()
    .a(43)
    .build();
```
You could also use a method-based builder and store some shared values in `self` to reuse them between calls. Basically imagine how you'd solve this problem without the builders using just functions with `Option<T>` parameters and then put a `#[builder]` on top of those functions.

**Comment by XBagon** [comment]
```rs
 let mut node_builder = GraphNodeBundle::builder(&node_assets, &shared_assets);
 for i in 0..100 {
    let node = node_builder.pos(Vec2::new(i as f32 * 5., 0.0)).build();
    commands.spawn(node.clone());
 }
```
This what I would like to currently, using it as a way to spawn multiple entities with different position in bevy.
A builder builder probably would work, but still would require cloning.

**Comment by Veetaha** [comment]
Do you think there is unnecessary cloning that can be avoided? If you are cloning the builder to create a new object that is composed of the values stores in the builder, then cloning  should be fine, because all the cloned values will be moved into the built object.

In the example you shown you call `node.clone()` (which seems unnecessary), and I see you are storing the builder in a mutable variable, plus the builder itself isn't cloned. That doesn't look like bon's builder pattern. I'd expect seeing a builder cloned or created from scratch (e.g. returned from a function).

**Comment by XBagon** [comment]
> Do you think there is unnecessary cloning that can be avoided? If you are cloning the builder to create a new object that is composed of the values stored in the builder, then cloning should be fine, because all the cloned values will be moved into the built object.

I haven't thought about that, yes I guess that should be optimized away then.
So then I can just express I prefer not having to write another builder and compiling another macro :p.

> In the example you shown you call node.clone() (which seems unnecessary), and I see you are storing the builder in a mutable variable, plus the builder itself isn't cloned. That doesn't look like bon's builder pattern. I'd expect seeing a builder cloned or created from scratch (e.g. returned from a function) on every iteration

Right I forgot to remove the clone!
The code snippet I share is what I would like to be able to do, not clone, instead modify the current builder and repeatedly "print" new items from the same builder which is updated, instead of recreated.

**Comment by Veetaha** [comment]
To simplify things, just imagine that your object consists of a single string field, and you are creating 100 objects that need to use the same string value, or maybe slightly modified version of it. Then you'll have to clone the string 100 times, because every object needs its own instance of the string. Storing the string in a `mut` variable won't solve the problem that you need 100 strings, which you can create either by cloning an existing string, or creating a brand new string on every iteration.

If you are concerned about having 100 clones of the same value, consider using a reference-counted pointer `Rc/Arc` to share the data instead of producing independent instances of it. If you reached for `bon` to work around the cloning problem, then yeah, it won't work around it, it will only give you nicer syntax for functions with many arguments and default values

**Comment by palapapa** [comment]
This is very useful if I want to have some helper functions that create partially configured builders to reduce code duplication. With overwritable builders, I don't have to create a whole new helper function just to not set a field I want to overwrite. I can share the same helper function for all cases and just overwrite some fields in some rare cases.

**Comment by Veetaha** [comment]
> This is very useful if I want to have some helper functions that create partially configured builders to reduce code duplication

Could you show an example code of this? I'm thinking what you'd like here can be expressed just via regular `#[builder(default)]` annotations.

> However, I have found a weird behavior.

Thanks for reporting that! I haven't looked into the root cause of the bug here, but it should definitely be fixed

**Comment by Sushisource** [comment]
> > > This is very useful if I want to have some helper functions that create partially configured builders to reduce code duplication
> > 
> > 
> > Could you show an example code of this? I'm thinking what you'd like here can be expressed just via regular `#[builder(default)]` annotations.
> > > However, I have found a weird behavior.
> > 
> > 
> > Thanks for reporting that! I haven't looked into the root cause of the bug here yet, but it should definitely be fixed
> 
> I am thinking that you mean I should just use `#[builder(default)]` to set the default value of some members? But I can't use that. The helper function will set the value of the members based on the arguments, and I want to allow the caller to overwrite the values set by the helper function if the user doesn't want the default.

I also have this exact use case. I don't necessarily need an `overwritable` attribute per-se. I have a handful of helpers in my code that set all the required fields for a builder to default values used for testing. These might vary by test environment or something - but what I do want is to be able to return `XXXBuilder<impl IsComplete>` and then be able to set optional fields on it that haven't been set - but as it stands today, you can't do that without just listing the whole concrete type of what is exactly set (because otherwise all fields are considered already-set). If there was some other trait I could use, like `impl IsOnlyNecessaryFields`, that is like `IsComplete` but _also_ with the bounds for all non-required fields set to `IsUnset`, that would help.

However, it still makes it difficult to then overwrite something the helper set that you _do_ want to change. Rather than annotating individual fields with `overwritable`, it might be nicer to have an `XXX::overwritable_builder()`, which allows all fields to be overwritten, maybe with the ability to flip to a normal builder by calling `seal()` or something.

Thanks so much for your work on this library!

**Comment by 0xangelo** [comment]
Just learned about this in the [reference](https://bon-rs.com/reference/builder/member/overwritable) and wanted to add that I definitely have a use for this similar to the "Dummy Values in Tests" section, so I'm looking forward to its stabilization!

Incredible crate BTW. I'm surprised I didn't learn about it earlier. I only found it because I decided to browse https://lib.rs/rust-patterns

**Comment by Veetaha** [comment]
@0xangelo, thank you! Also, take a look at this comment. I think it may be useful as an alternative to using the `overwritable` feature for dummy values in tests https://github.com/elastio/bon/issues/149#issuecomment-3618598357

**Comment by 0xangelo** [comment]
Makes sense! I started using "wrapper builder functions" as in your first suggestion and it covers my case pretty well.

There's some repetition, but it's better that what I was doing before I think because the builder was defined in another crate so I was making the typestate module `pub` and in my integration tests I was returning something like
```rust
pub fn my_test_setup_helper(/* ... */) -> SomeBuilder<
    impl State<
        FieldA: IsSet,
        FieldB: IsSet,
        FieldC: IsUnset,
    >
>
```
I worked but I was exposing too much in hindsight
